src/cls/rgw/cls_rgw.cc

```diff
@@ -809,17 +809,27 @@ static int read_key_entry(cls_method_context_t hctx, cls_rgw_obj_key& key,
   return 0;
 }
 
-int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
-{
-  // decode request
-  rgw_cls_obj_complete_op op;
-  auto iter = in->cbegin();
+template <typename T>
+static int decode_input_op(bufferlist *in, T& op){
```

Comment:
prefer bufferlist* (or buffer::list*) for new code if we can

Reply:
ack